### DISCUSSION TITLE
How is everyone bootstrapping with GitHub?

### DISCUSSION THREAD

**Discussion by cunninghamj-reninc** [fluxcd/flux2#4163] [Q&A]
I am trying to make a GitHub Actions Workflow that will run `flux bootstrap github`, but am not having much luck. GitHub Workflows have a predefined environment variable `GITHUB_TOKEN`. This is fine, but flux is also looking for an environment variable named that. The problem is that the token that is generated for the workflow does not have the right permissions and this environment variable's value can not be overwritten to that of a PAT. Right now I am seeing this error:

```shell
GET https://api.github.com/repos/<redacted>/<redacted>/keys: 403 Resource not accessible by personal access token []
```

Is there something I am missing? Or is everyone bootstrapping their clusters a different way?

**Comment by raffraffraff** [comment]
It's surprising that nobody answered this after 8 months, since it happens kinda 'by default'. 

Q. Why is FluxCD trying to create a key? What sort of key is it?
A. FluxCD wants to create a repository-specific read-only _deploy key_ (basically, an SSH key). It does by default, because fine-grained access tokens typically belong to _users_, and if if you're working on a team, they won't want some important project to rely on your personal access token!

Q. What permission should I give to the fine-grained access token so that it can generate the key?
A. This isn't clearly documented anywhere, but you need restrict the token to a specific repo, then grant it Repository > Administration > Read and Write (in addition to the other rights it needs, like Content, Metadata etc)

Q. OK, I'm cool with using a deploy key.. but if FluxCD creates a read-only key, how can it ever update the git project again?
A. It can't, and that's the point. During the bootstrap process, FluxCD requires write access so that it can push the cluster config back to git, but after that all of your changes should be created outside the cluster, so it can operate in pull mode. See this:  https://github.com/fluxcd/flux2/discussions/1819#discussioncomment-1322812

Q. What if I don't want it to create a deploy key? What if I just want it to your my fine-grained access token?
A. *I think* you can bootstrap with the `--token-auth` option, which instructs FluxCD to use your fine-grained PAT instead of the deploy key.

Ultimately the security choices are:
1. Give FluxCD its own Github account, and in it create a fine-grained personal access token. Bootstrap with using `--token-auth`.
2. Bootstrap FluxCD with your own personal access token, and let it generate a repo-specific (default: read-only) deploy key
